[Concurrency] Adopt typed throws and nonisolated(nonsending) in withTaskCancellationHandler

DougGregor opened the pull request:
Replace the use of rethrows and #isolation in
withTaskCancellationHandler with typed throws and
nonisolated(nonsending), respectively.

Fixes rdar://146901428.

DougGregor committed:
[Concurrency] Adopt typed throws and nonisolated(nonsending) in withTaskCancellationHandler

Replace the use of rethrows and #isolation in
withTaskCancellationHandler with typed throws and
nonisolated(nonsending), respectively.

Fixes rdar://146901428.

DougGregor committed:
Drop extraneous `isolation` argument

DougGregor commented:
@swift-ci please smoke test

ktoso commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -74,10 +74,29 @@ import Swift
+public func withTaskCancellationHandler<T, E>(
+  operation: () async throws(E) -> T,
+  onCancel handler: @Sendable () -> Void
+) async throws(E) -> T {
```
This has been a long standing bug. We always "promised" that these with methods would never lose isolation however this didn't actually work correctly with the isolated parameter... 

This is finally bringing the correct behavior to these APIs. All the with methods have the same issue and we'll change all of them to now give the behavior we've always promised to begin with

ktoso replied:
Well, that is the bug:

```
    await withTaskCancellationHandler {
        MainActor.assertIsolated() // fails!
```

this is very much incorrect and should have never been happening to hop off, but it does. The way closure isolation works just isn't correct here and the "real" solution is to move all these with... APIs to nonisolated nonsending.

> it seems like this change alone (ignoring typed throws) will not alter the behavior of this API 

It does change behavior; it corrects the unexpected hopping off in certain scenarios -- one of which you've just pasted, that is incorrect and we must fix it; it breaks the safety contract of the withTaskCancellationHandler function

DougGregor commented:
https://github.com/swiftlang/swift-foundation/pull/1602

@swift-ci please smoke test Linux

DougGregor commented:
https://github.com/swiftlang/swift-docc/pull/1351

@swift-ci please smoke test Windows

DougGregor commented:
https://github.com/swiftlang/swift-docc/pull/1351

@swift-ci please smoke test Windows

DougGregor commented:
https://github.com/swiftlang/swift-docc/pull/1351

@swift-ci please smoke test Windows

DougGregor commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -74,10 +74,29 @@ import Swift
+@export(implementation)
+nonisolated(nonsending)
+public func withTaskCancellationHandler<T, E>(
+  operation: () async throws(E) -> T,
```
Good catch, thank you!

DougGregor commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -74,10 +74,29 @@ import Swift
+nonisolated(nonsending)
+public func withTaskCancellationHandler<T, E>(
+  operation: () async throws(E) -> T,
+  onCancel handler: @Sendable () -> Void
```
That's a good point!

DougGregor committed:
[Task cancellation] Adopt nonisolated(nonsending) and sending

ktoso commented:
@swift-ci please smoke test

ktoso commented:
This is a part of https://github.com/swiftlang/swift/pull/80753 which does it for all APIs but is having trouble landing, so it would be fine to just land this one instead and start addressing the methods one by one.

cc @xedin

xedin committed:
[stdlib] Replace `@export(implementation)` with `@_AEIC` on `withTaskCancellationHandler`

`@export(implementation)` cannot be widely adopted in stdlib yet,
currently leads to condfail in some configrations. Stick to `@_alwaysEmitIntoClient`
for now.

ktoso commented:
@swift-ci please smoke test
